field deregister bug +  dynamic array of scopes

glebtv opened the pull request:
1) This fixes field deregister bug, which did not mark the field as not registered on deregister.
This was required for this, so after removing an item from array and re-adding it back, initial value is re-read.
2) Adds an array of dynamic scopes component (aka nested form) and storybook demo for it.

glebtv commented:
Is ```savedApi.setValue("siblings[0].name", "Test")``` the expected nested field syntax?
Maybe it should be documented somewhere?
```savedApi.setValue("siblings[0][name]", "Test")``` does not work for me

joepuzzo commented:
Taking a look now.. honestly one of the best PRs i have seen ( you added tests and docs wooo! ). Im still going through the code but so far i have noticed that add with preset is presetting all new fields. To recreate use `Add` button to add a field, then add with preset.

joepuzzo commented on src/components/ScopeArrayField.js:
```diff
@@ -0,0 +1,75 @@
+  var initialValues = formApi.getInitialValue(field)
+
+  if (!initialValues) {
+    throw(new Error("initialValue required for ScopeArrayField " + field))
```
Why are initial values required?

joepuzzo commented on src/components/ScopeArrayField.js:
```diff
@@ -0,0 +1,75 @@
+    throw(new Error("initialValue should be an array for ScopeArrayField " + field))
+  }
+
+  const initialKeys = initialValues.map((v, index) => index);
```
If initial values are not required ( which i believe they should NOT be ) then this will need a null check like this line in Array Field https://github.com/joepuzzo/informed/blob/master/src/components/ArrayField.js#L21

joepuzzo commented on src/FormController.js:
```diff
@@ -374,6 +374,7 @@ class FormController extends EventEmitter {
       ObjectMap.delete(this.state.values, field);
       ObjectMap.delete(this.state.touched, field);
       ObjectMap.delete(this.state.errors, field);
+      this.registered[field] = false;
```
So I think you have a misunderstanding of what `this.registered` is. That is an object that remembers if a field has been registered ever! not wether or not its currently registered. By adding this you broke Array field because when you remove an element in the middle of the array it will not have the correct value.

joepuzzo replied:
https://github.com/joepuzzo/informed/blob/master/src/FormController.js#L358 <<< this line is where it sets the value to the old value if the field was "once registered"

joepuzzo commented on stories/Arrays/DynamicArrayOfScopes/README.md:
```diff
@@ -0,0 +1,75 @@
+                        )
+                      })}
+
+                      <button onClick={() => {
```
Make sure these buttons are `type="button"` If you dont they will have the default submit type and submit the form!

joepuzzo commented:
@glebtv So i'm not understanding the need for the scoped array field? I just updated your storybook story to use the normal `ArrayField` and it still works fine...

joepuzzo commented on stories/Arrays/DynamicArrayOfScopes/README.md:
```diff
@@ -0,0 +1,75 @@
+    >
+      {({ formApi, formState }) => {
+        return (
+          <div style={{ display: 'flex', flexWrap: 'wrap' }}>
```
If you look at all the other docs.. i never show the flex stuff and how the state is rendered with the `FormState` comp. The REAMEs are always simplified for a reason, this should follow that pattern

joepuzzo commented:
@glebtv OK SO THIS IS THE MOST IMPORTANT COMMENT! YOU DID NOT NEED TO CREATE SCOPEDARRAYFIELD! 

```
import React from 'react';
import withDocs from '../../utils/withDocs';
import readme from './README.md';

import FormState from '../../utils/FormState';
import Code from '../../utils/Code';

import { Form, Text, ArrayField } from '../../../src';

const initialValues = {
  name: 'test',
  friends: [{
    name: 'Joe',
    age: '20',
  },{
    name: 'Jane',
    age: '20',
  }],
};

const DynamicArrayOfScopes = () => (
  <div>
    <Form
      initialValues={initialValues}
    >
      {({ formApi, formState }) => {
        return (
          <div style={{ display: 'flex', flexWrap: 'wrap' }}>
            <div style={{ flex: 1, marginRight: '2rem' }}>
              <Text field="name"/>

              <ArrayField field="friends">
                {({ add, fields }) => {
                  return (
                    <React.Fragment>
                      {fields.map(({key, field, remove}) => {
                        return (
                          <label key={key}>
                            <h5>{field}</h5>
                            <Text field={`${field}.name`}/>
                            <Text field={`${field}.age`}/>
                            <button onClick={remove}>Remove</button>
                          </label>
                        );
                      })}

                      <button onClick={() => {
                        add();
                      }} type="button">Add</button>

                      <button onClick={() => {
                        add({name: 'test'});
                      }}>Add with preset</button>

                      <button onClick={() => {
                        formApi.setValue('friends[0].name', 'Test');
                      }} type="button">set friends[0].name to test</button>
                    </React.Fragment>
                  );

                }}
              </ArrayField>
            </div>
            <div style={{ flex: 2, minWidth: '300px' }}>
              <FormState />
            </div>
          </div>
        );
      }}

    </Form>
  </div>
);

export default withDocs(readme, DynamicArrayOfScopes);

```

glebtv commented:
Thank you for the feedback,

Honestly, this is like my 10th try to make an array of scopes with add/delete work properly, it turned out to be extremely hard to do :)

I will take a look and try to make it work with ArrayField and update this PR with the storybooks and tests.

joepuzzo commented:
Yeah no need to try to make it work.. literally just use that code snippit i sent with the current informed source code and it works

glebtv commented:
I need an ability to add an item to array with some fields pre-set (for me it's kind, id, etc),

this was very hard without changing something in informed as I can't do setValue before informed rendered the fields;

there is no such thing in ArrayField

Are you okay if I add an ability to add item with a pre-set value to ArrayField?

glebtv commented:
More specifically,
```
                     <button type="button" onClick={() => {
                       var newFriends = formState.values.friends || [];
                       var maxId = 0
                       newFriends.forEach((v) => {
                         if (v.id > maxId) {
                           maxId = v.id
                         }
                       })
                       console.log("maxid", maxId)
                       add()
                       formApi.setValue(`friends.${newFriends.length}`, {id: maxId + 1})
                      }}>Add</button>
```

This causes:
```
FormController.js:74 Uncaught TypeError: Cannot read property 'fieldApi' of undefined
    at Object.setValue (FormController.js:74)
    at onClick (index.js:61)
```

glebtv commented:
It works if I wrap the ```setValue``` in ```setTimeout(() => ,0)``` - but for some reason, not all the time and not for two fields, for example for me this sets id, but name fails with that error.
```
                       add()
                       setTimeout(() => {
                         formApi.setValue(`friends[${newFriends.length}].id`, maxId + 1)
                         formApi.setValue(`friends[${newFriends.length}].name`, "test")
                       },0)
```

joepuzzo commented:
Ok i can probably just look into adding the ability to add with an initial value

glebtv commented:
I've changed my examples and tests to use ArrayField as you suggested here:
https://github.com/glebtv/informed/tree/array_v2
https://github.com/joepuzzo/informed/compare/master...glebtv:array_v2?expand=1

Also fixed most of your comments about examples

For me it works poorly,
1) adding an item with preset sets content for the last empty item, not for the intended newly-added last array item
2) removing item sometimes fails to clear formState - see failing test

I think ArrayField is broken since there should be no difference between removed field and a field that never existed..

And yep, my code relies on mutating InitialState array - i'll say that explicitly since I think you did not notice..

glebtv commented:
Using a separate addWithValue because ```<button onClick={add}>``` breaks if I add a parameter (since it will receive an event)

joepuzzo commented:
So there is a difference between a removed field and a field that never existed and here is why: 

Suppose we have three inputs for friends

-----------------------
|              Input1             |
-----------------------

-----------------------
|              Input2             |
-----------------------

-----------------------
|              Input3             |
-----------------------

This would be represented in the form state as follows: 

```
{
  friends: [
    input1 = friends[0], 
    input2 = friends[1],
    input3 = friends[2]
  ]
}
```

NOW suppose we remove `input2`

What happens?? 

1. the field friends[2] NO LONGER EXISTS. Therefore it performs a `deregister`. 

2. the field name for input3 changes from `friends[2]` to become `friends[1]` Therefore input3's effect gets triggered and it gets deregistered as  `friends[2] and reregistered at `friends[1]` 

So whats the point of all this? It shows a scenario where a field that was removed but added back is different than a field that never existed.

joepuzzo commented:
Let me take a stab at adding with a value and get back to you

joepuzzo commented:
I did it! Stand by

joepuzzo commented:
Ok so i made a PR on your fork with my changes. Take a look and lmk what you think!

glebtv commented:
For me this now works really nice both in the demo and in my app, and resolves all the problems I had implementing a nested form.

Thank you 👍

glebtv commented:
Considering field deregistration - honestly, I still don't get why it removing/adding back a field makes it different.

1) Initial:
```
{
  friends: [
    input1 = friends[0], 
    input2 = friends[1],
    input3 = friends[2]
  ]
}
```

2) remove field 2 (input2):
```
{
  friends: [
    input1 = friends[0], 
    input3 = friends[2]
  ]
}
```

3) field 3 deregisters, re-registers as field 2, form state should be updated to remove old field2's value from the array (friends[2] becomes friends[1]):
```
{
  friends: [
    input1 = friends[0], 
    input2 = friends[1] 
  ]
}
```

4) Add the field back, everything is exactly the same as before removing a field:
```
{
  friends: [
    input1 = friends[0], 
    input2 = friends[1],
    input3 = friends[2]
  ]
}
```

glebtv commented:
I think I get it now - removing this flag makes it re-read initialValues, which is not the correct behaviour for deleted and re-added field.

Thank you again :)

glebtv commented:
But in fact this is the only difference - it works without manually passing down ```initialValues``` for non-pre-exising fields in your version, so if you remove the flag on deregister, it will work without passing down ```initialValues```  as mine did.

In ```addWithInitialValue``` you also mutate pre-existing initialValues, same as mine, and that's what causes it to appear in the form (is no ```setInitialValues``` in ```addWithInitialValue``` intentional?)

joepuzzo commented:
ooo that may be a mistake stand by

joepuzzo commented:
```
const addWithInitialValue = ( initialValue ) => {
    keys.push(uuidv4());
    setKeys([...keys]);
    const newInitialValues = [...initialValues];
    newInitialValues[keys.length - 1] = initialValue;
    setInitialValues(newInitialValues);
  };
```

Im going to change it to this before i merge

joepuzzo commented:
Or rather this

```
const addWithInitialValue = ( initialValue ) => {
    keys.push(uuidv4());
    setKeys([...keys]);
    initialValues[keys.length - 1] = initialValue;
    setInitialValues([...initialValues]);
  };
```

glebtv commented:
Your second one still mutates initial values :)
This seems to be right

joepuzzo commented:
Ok cool yeah.. I'm going to review and merge tomorrow its getting late where i am so i want to make sure its all good before i release.

joepuzzo commented:
Found a bug with addWithInital value and your implementation. Updating now and will release in 2.8.1
